pkg/controller/batch/batch_controller.go

```diff
@@ -0,0 +1,392 @@
+			return controllerutil.SetControllerReference(batch, configMap, r.scheme)
+		})
+
+		if err != nil {
+			return reconcile.Result{}, fmt.Errorf("Unable to create ConfigMap '%s': %w", configMap.Name, err)
+		}
+
+		updated, err := r.update(func() {
+			batch.Spec.ConfigMap = &batch.Name
+		})
+		if err != nil {
+			return reconcile.Result{}, fmt.Errorf("Unable to update Batch .Spec: %w", err)
+		}
+		if updated {
+			return reconcile.Result{}, nil
+		}
+	}
+
+	// We update the phase separately to the spec as the status update is ignored when in the update mutate function
+	return reconcile.Result{}, r.UpdatePhase(v2.BatchInitialized, nil)
+}
+
+func (r *batchResource) execute() (reconcile.Result, error) {
+	infinispan, err := r.loadCluster()
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	batch := r.instance
+	url, err := connectionUrl(infinispan)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	cliArgs := fmt.Sprintf("--connect '%s' --file '%s/%s'", url, BatchVolumeRoot, BatchFilename)
+
+	labels := batchLabels(batch.Name)
+	infinispan.AddLabelsForPods(labels)
+
+	job := &batchv1.Job{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      batch.Name,
+			Namespace: batch.Namespace,
+		},
+		Spec: batchv1.JobSpec{
+			BackoffLimit: pointer.Int32Ptr(0),
+			Template: corev1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: labels,
+				},
+				Spec: corev1.PodSpec{
+					Containers: []corev1.Container{{
+						Name:    batch.Name,
+						Image:   infinispan.ImageName(),
+						Command: []string{Entrypoint, cliArgs},
+						VolumeMounts: []corev1.VolumeMount{
+							{
+								Name:      BatchVolumeName,
+								MountPath: BatchVolumeRoot,
+							},
+						},
+					}},
+					RestartPolicy: corev1.RestartPolicyNever,
+					Volumes: []corev1.Volume{{
+						Name: BatchVolumeName,
+						VolumeSource: corev1.VolumeSource{
+							ConfigMap: &corev1.ConfigMapVolumeSource{
+								LocalObjectReference: corev1.LocalObjectReference{Name: batch.Name},
+							},
+						},
+					}},
+				},
+			},
+		},
+	}
+
+	_, err = controllerutil.CreateOrUpdate(ctx, r.client, job, func() error {
+		return controllerutil.SetControllerReference(batch, job, r.scheme)
+	})
+
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("Unable to create batch job '%s': %w", batch.Name, err)
+	}
+	return reconcile.Result{}, r.UpdatePhase(v2.BatchRunning, nil)
+}
+
+func (r *batchResource) waitToComplete() (reconcile.Result, error) {
+	batch := r.instance
+	job := &batchv1.Job{}
+	jobKey := types.NamespacedName{
+		Name:      batch.Name,
+		Namespace: batch.Namespace,
+	}
+	if err := r.client.Get(ctx, jobKey, job); err != nil {
+		if errors.IsNotFound(err) {
+			return reconcile.Result{}, fmt.Errorf("Job '%s' not found", jobKey.Name)
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, fmt.Errorf("Unable to fetch Job '%s': %w", jobKey.Name, err)
+	}
```

Comment:
We have a `LookupResource` function in infinispan_controller for this case. NotFound means not ready atm.

Reply:
Updated and I have moved `LookupResource` to `pkg/kubernertes/controllerutil.go` as this is a generic method applicable to all of our controllers.